[visionOS] Repeated attempts to enter and exit LinearMediaPlayer fullscreen may fail

aestes opened the pull request:
#### f1122551722bf80eedb0eaf548808a999282e0ce
<pre>
[visionOS] Repeated attempts to enter and exit LinearMediaPlayer fullscreen may fail
<a href="https://bugs.webkit.org/show_bug.cgi?id=272253">https://bugs.webkit.org/show_bug.cgi?id=272253</a>
<a href="https://rdar.apple.com/125994291">rdar://125994291</a>

Reviewed by Jer Noble.

When exiting LinearMediaPlayer fullscreen LinearMediaKit may call willExitFullscreen() twice, and
each time PlaybackSessionInterfaceLMK would respond by calling
PlaybackSessionModel::toggleFullscreen(). The first call would exit fullscreen and ultimately
destroy the LinearMediaPlayer, but since the second call would occur before
VideoPresentationInterfaceIOS::cleanupFullscreen() was called the fullscreen state machine would be
left in a bad state where it believed it was in standard fullscreen but a LinearMediaPlayer would
not exist. Later attempts to enter fullscreen would fail due to the LMPlayableViewController not
having a valid Playable-conforming object.

Resolved this by teaching LinearMediaPlayer to track a presentationState indepenently of the value
of presentationMode, allowing it to tolerate multiple calls to willEnterFullscreen(),
willExitFullscreen(), and toggleInlineMode(). When LinearMediaKit calls willEnterFullscreen() and
presentationState is .inline we change it to .enteringFullscreen an ddispatch
-linearMediaPlayerEnterFullscreen:, otherwise we ignore the call. When LinearMediaKit calls
willExitFullscreen() and presentationState is .fullscreen we change it to .exitingFullscreen and
dispatch -linearMediaPlayerExitFullscreen:, otherwise we ignore the call. When LinearMediaKit calls
toggleInlineMode() and presentationState is .inline or .fullscreen, we change it to
.enteringFullscreen or .exitingFullscreen respectively, otherwise we ignore the call.

Next, rather than allowing LinearMediaPlayer&apos;s client to mutate presentationMode directly, exposed
-enterFullscreen and -exitFullscreen methods. When -enterFullscreen is called when presentationState
is anything but .fullscreen, we change presentationState to .fullscreen and presentationMode to
.fullscreenFromInline. When -exitFullscreen is called when presentationState is anything but
.inline, we change presentationState and presentationMode to .inline.

Finally, PlaybackSessionModel::toggleFullscreen() was removed and an explicit
PlaybackSessionModel::exitFullscreen() was added. When PlaybackSessionInterfaceLMK receives
-linearMediaPlayerEnterFullscreen: or -linearMediaPlayerExitFullscreen: it calls
PlaybackSessionModel::enterFullscreen() and PlaybackSessionModel::exitFullscreen() respectively.
VideoPresentationInterfaceLMK was also changed to call -enterFullscreen and -exitFullscreen instead
of mutating LinearMediaPlayer&apos;s presentationMode directly.

The net result of these changes is that a request from LinearMediaKit to enter or exit fullscreen is
ignored if a previous request is still in progress or if the state machine is already in the
requested state.

* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::exitFullscreen):
(WebCore::PlaybackSessionModelMediaElement::toggleFullscreen): Deleted.
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
(-[WKLinearMediaPlayerDelegate linearMediaPlayerEnterFullscreen:]):
(-[WKLinearMediaPlayerDelegate linearMediaPlayerExitFullscreen:]):
(-[WKLinearMediaPlayerDelegate linearMediaPlayerToggleInlineMode:]): Deleted.
(-[WKLinearMediaPlayerDelegate linearMediaPlayerWillEnterFullscreen:]): Deleted.
(-[WKLinearMediaPlayerDelegate linearMediaPlayerWillExitFullscreen:]): Deleted.
(-[WKLinearMediaPlayerDelegate _exitFullscreen]): Deleted.
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
(WebKit::VideoPresentationInterfaceLMK::setupPlayerViewController):
(WebKit::VideoPresentationInterfaceLMK::presentFullscreen):
(WebKit::VideoPresentationInterfaceLMK::dismissFullscreen):
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::exitFullscreen):
(WebKit::PlaybackSessionManagerProxy::exitFullscreen):
(WebKit::PlaybackSessionModelContext::toggleFullscreen): Deleted.
(WebKit::PlaybackSessionManagerProxy::toggleFullscreen): Deleted.
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
(SwiftOnlyData.presentationMode):
(SwiftOnlyData.presentationState):
(WKSLinearMediaPlayer.presentationState):
(WKSLinearMediaPlayer.cancellables):
(WKSLinearMediaPlayer.enterFullscreen):
(WKSLinearMediaPlayer.exitFullscreen):
(WKSLinearMediaPlayer.presentationStateChanged(_:)):
(WKSLinearMediaPlayer.presentationModePublisher):
(WKSLinearMediaPlayer.toggleInlineMode):
(WKSLinearMediaPlayer.willEnterFullscreen):
(WKSLinearMediaPlayer.willExitFullscreen):
(WKSLinearMediaPlayer.presentationMode): Deleted.
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
(WKSLinearMediaPresentationMode.presentationMode): Deleted.
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaPlayer.h:
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaTypes.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.messages.in:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionManager::exitFullscreen):
(WebKit::PlaybackSessionManager::toggleFullscreen): Deleted.

Canonical link: <a href="https://commits.webkit.org/277169@main">https://commits.webkit.org/277169@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ad74a8b58c6f66db82bcb09bce5e877d9a451afb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46882 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/26045 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49509 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49562 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42930 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/49189 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30693 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23511 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38174 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47463 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/23190 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40376 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19484 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20523 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41516 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4928 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43142 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41892 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51436 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21894 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18239 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45469 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23179 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44442 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10352 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/23769 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22890 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
```diff
@@ -114,7 +114,7 @@ -(void)didFinishFullscreen:(VideoFullscreenControllerContext*)context;
     void setController(WebVideoFullscreenController* controller) { m_controller = controller; }
     void setUpFullscreen(HTMLVideoElement&, UIView *, HTMLMediaElementEnums::VideoFullscreenMode);
-    void exitFullscreen();
+    void exitFullscreen() override;
```
Can't this be `final`?

aestes committed:
[visionOS] Repeated attempts to enter and exit LinearMediaPlayer fullscreen may fail
https://bugs.webkit.org/show_bug.cgi?id=272253
rdar://125994291

Reviewed by Jer Noble.

When exiting LinearMediaPlayer fullscreen LinearMediaKit may call willExitFullscreen() twice, and
each time PlaybackSessionInterfaceLMK would respond by calling
PlaybackSessionModel::toggleFullscreen(). The first call would exit fullscreen and ultimately
destroy the LinearMediaPlayer, but since the second call would occur before
VideoPresentationInterfaceIOS::cleanupFullscreen() was called the fullscreen state machine would be
left in a bad state where it believed it was in standard fullscreen but a LinearMediaPlayer would
not exist. Later attempts to enter fullscreen would fail due to the LMPlayableViewController not
having a valid Playable-conforming object.

Resolved this by teaching LinearMediaPlayer to track a presentationState indepenently of the value
of presentationMode, allowing it to tolerate multiple calls to willEnterFullscreen(),
willExitFullscreen(), and toggleInlineMode(). When LinearMediaKit calls willEnterFullscreen() and
presentationState is .inline we change it to .enteringFullscreen an ddispatch
-linearMediaPlayerEnterFullscreen:, otherwise we ignore the call. When LinearMediaKit calls
willExitFullscreen() and presentationState is .fullscreen we change it to .exitingFullscreen and
dispatch -linearMediaPlayerExitFullscreen:, otherwise we ignore the call. When LinearMediaKit calls
toggleInlineMode() and presentationState is .inline or .fullscreen, we change it to
.enteringFullscreen or .exitingFullscreen respectively, otherwise we ignore the call.

Next, rather than allowing LinearMediaPlayer's client to mutate presentationMode directly, exposed
-enterFullscreen and -exitFullscreen methods. When -enterFullscreen is called when presentationState
is anything but .fullscreen, we change presentationState to .fullscreen and presentationMode to
.fullscreenFromInline. When -exitFullscreen is called when presentationState is anything but
.inline, we change presentationState and presentationMode to .inline.

Finally, PlaybackSessionModel::toggleFullscreen() was removed and an explicit
PlaybackSessionModel::exitFullscreen() was added. When PlaybackSessionInterfaceLMK receives
-linearMediaPlayerEnterFullscreen: or -linearMediaPlayerExitFullscreen: it calls
PlaybackSessionModel::enterFullscreen() and PlaybackSessionModel::exitFullscreen() respectively.
VideoPresentationInterfaceLMK was also changed to call -enterFullscreen and -exitFullscreen instead
of mutating LinearMediaPlayer's presentationMode directly.

The net result of these changes is that a request from LinearMediaKit to enter or exit fullscreen is
ignored if a previous request is still in progress or if the state machine is already in the
requested state.

* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::exitFullscreen):
(WebCore::PlaybackSessionModelMediaElement::toggleFullscreen): Deleted.
* Source/WebCore/platform/ios/WebVideoFullscreenControllerAVKit.mm:
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
(-[WKLinearMediaPlayerDelegate linearMediaPlayerEnterFullscreen:]):
(-[WKLinearMediaPlayerDelegate linearMediaPlayerExitFullscreen:]):
(-[WKLinearMediaPlayerDelegate linearMediaPlayerToggleInlineMode:]): Deleted.
(-[WKLinearMediaPlayerDelegate linearMediaPlayerWillEnterFullscreen:]): Deleted.
(-[WKLinearMediaPlayerDelegate linearMediaPlayerWillExitFullscreen:]): Deleted.
(-[WKLinearMediaPlayerDelegate _exitFullscreen]): Deleted.
* Source/WebKit/Platform/ios/VideoPresentationInterfaceLMK.mm:
(WebKit::VideoPresentationInterfaceLMK::setupPlayerViewController):
(WebKit::VideoPresentationInterfaceLMK::presentFullscreen):
(WebKit::VideoPresentationInterfaceLMK::dismissFullscreen):
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::exitFullscreen):
(WebKit::PlaybackSessionManagerProxy::exitFullscreen):
(WebKit::PlaybackSessionModelContext::toggleFullscreen): Deleted.
(WebKit::PlaybackSessionManagerProxy::toggleFullscreen): Deleted.
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaPlayer.swift:
(SwiftOnlyData.presentationMode):
(SwiftOnlyData.presentationState):
(WKSLinearMediaPlayer.presentationState):
(WKSLinearMediaPlayer.cancellables):
(WKSLinearMediaPlayer.enterFullscreen):
(WKSLinearMediaPlayer.exitFullscreen):
(WKSLinearMediaPlayer.presentationStateChanged(_:)):
(WKSLinearMediaPlayer.presentationModePublisher):
(WKSLinearMediaPlayer.toggleInlineMode):
(WKSLinearMediaPlayer.willEnterFullscreen):
(WKSLinearMediaPlayer.willExitFullscreen):
(WKSLinearMediaPlayer.presentationMode): Deleted.
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
(WKSLinearMediaPresentationMode.presentationMode): Deleted.
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaPlayer.h:
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaTypes.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.messages.in:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionManager::exitFullscreen):
(WebKit::PlaybackSessionManager::toggleFullscreen): Deleted.

Canonical link: https://commits.webkit.org/277169@main
